File: tests/python/unit/test_unit_dataset.py
from dataset import AudioDataset
import pytest
import torch

@pytest.fixture(params=[True, False])
def preload(request):
  return request.param

@pytest.fixture
def mock_dataset(tmp_path):
  def _make_dataset(preload: bool):
    root_path = tmp_path / "resources"
    root_path.mkdir(parents=True, exist_ok=True)

    if not preload:
      genres = ["jazz", "rock"]
      for genre in genres:
        dir_path = root_path / genre
        dir_path.mkdir(parents=True, exist_ok=True)

        for i in range(3):
          file_path = dir_path / f"{genre}.0000{i}.wav"
          file_path.write_bytes(b"Mock...Data")
    else:
      dataset_path = root_path / "dataset_py.pt"
      dataset = (
        [torch.rand((1290, 13)) for _ in range(6)],
        [torch.rand(1) for _ in range(6)],
        {0: "jazz", 1: "rock"}
      )
      torch.save(dataset, dataset_path)
    return root_path
  return _make_dataset

@pytest.fixture
def audio_dataset_with_params(preload, mock_dataset, mocker):
  root_path = mock_dataset(preload=preload)

  mock_preprocessor = mocker.Mock()
  mock_preprocessor.process_file.side_effect = lambda _: torch.rand((1290, 13))
  mock_preprocessor.normalize_data.side_effect = lambda _: torch.rand((1290, 13))
  return AudioDataset(root_path, mock_preprocessor)

def test_dataset_getitem(audio_dataset_with_params):
  num_frames = 1290
  mfcc_size = 13

  data, target = audio_dataset_with_params[0]
  assert(data.numel() == num_frames * mfcc_size), "Invalid size of the samle data."
  assert(target.numel() == 1), "Invalid size of the sample target."

@pytest.fixture
def audio_dataset(mock_dataset):
  root_path = mock_dataset(preload=True)
  return AudioDataset(root_path, None)

def test_dataset_classes(audio_dataset):
  classes_expected = {0: "jazz", 1: "rock"}
  classes = audio_dataset.classes
  assert(classes == classes_expected), "Invalid dataset classes."

def test_dataset_data(audio_dataset):
  size_expected = 6
  num_frames = 1290
  mfcc_size = 13

  size = len(audio_dataset.data)
  data = audio_dataset.data[0]
  assert(size == size_expected), "Invalid size of the dataset."
  assert(data.numel() == num_frames * mfcc_size), "Invalid size of the samle data."

def test_dataset_len(audio_dataset):
  size_expected = 6
  size = len(audio_dataset)
  assert(size == size_expected), "Invalid size of the dataset."

File: tests/cpp/unit/TestUnitDataset.cpp
#include "Dataset.h"
#include <fstream>
#include <gmock/gmock.h>
#include <gtest/gtest.h>

using ::testing::_;
using ::testing::Return;
using ::testing::Throw;

struct MockDataset
{
  static void Mock(fs::path& rootPath, bool preload, size_t size, long H, long W)
  {
    rootPath = fs::temp_directory_path() / "resources";
    fs::create_directory(rootPath);

    if (!preload)
    {
      std::vector<std::string> genres = {"jazz", "rock"};
      for (const auto& genre : genres)
      {
        fs::path dirPath = rootPath / genre;
        fs::create_directory(dirPath);

        for (int i = 0; i < size / genres.size(); ++i)
        {
          fs::path filePath = dirPath / (genre + ".0000" + std::to_string(i) + ".wav");
          std::ofstream file(filePath);
          file << "Mock...Data";
        }
      }
    }
    else
    {
      std::vector<torch::Tensor> data(size, torch::rand({H, W}));
      std::vector<torch::Tensor> target(size, torch::rand(1));
      c10::Dict<int64_t, std::string> classes;
      classes.insert(0, "jazz");
      classes.insert(1, "rock");

      fs::path datasetPath = rootPath / "dataset_cpp.pt";
      torch::serialize::OutputArchive dataset;
      dataset.write("data", data);
      dataset.write("target", target);
      dataset.write("classes", classes);
      dataset.save_to(datasetPath);
    }
  }
};

class MockPreprocessor : public Preprocessor
{
  public:
    MockPreprocessor(PreprocessorConfig cfg = PreprocessorConfig()) : Preprocessor(cfg) {}
    MOCK_METHOD(torch::Tensor, processFile, (const fs::path& filePath), (override));
    MOCK_METHOD(torch::Tensor, normalizeData, (const torch::Tensor& x), (override));
};

class TestInitializationParam : public ::testing::TestWithParam<std::string>
{
  protected:
    fs::path rootPath;
    long H, W;
    size_t datasetSize;
    AudioDataset* audioDataset = nullptr;
    MockPreprocessor* mockPreprocessor = nullptr;

    void SetUp() override
    {
      std::string init = GetParam();
      datasetSize = 6;
      H = 1290;
      W = 13;
      MockDataset::Mock(rootPath, (init == "preload") ? true : false, datasetSize, H, W);
      mockPreprocessor = new MockPreprocessor();

      if (init == "corrupt")
      {
        EXPECT_CALL(*mockPreprocessor, processFile(_))
          .WillOnce(Throw(std::runtime_error("Preprocessor::loadAndCrop: Invalid or corrupted file.")))
          .WillRepeatedly(Return(torch::rand({1, H, W})));
        ASSERT_THROW(audioDataset = new AudioDataset(rootPath, mockPreprocessor), std::runtime_error);
      }
      else if (init == "init")
      {
        EXPECT_CALL(*mockPreprocessor, processFile(_))
          .WillRepeatedly(Return(torch::rand({1, H, W})));
      }

      audioDataset = new AudioDataset(rootPath, mockPreprocessor);
    }

    void TearDown() override
    {
      fs::remove_all(rootPath);
      delete mockPreprocessor;
      mockPreprocessor = nullptr;
      delete audioDataset;
      audioDataset = nullptr;
    }
};

TEST_P(TestInitializationParam, get)
{
  EXPECT_CALL(*mockPreprocessor, normalizeData(_)).WillOnce(Return(torch::rand({1, H, W})));
  torch::data::Example<> sample = audioDataset->get(0);
  ASSERT_EQ(sample.data.numel(), H * W) << "Invalid size of the sample data.";
  ASSERT_EQ(sample.target.numel(), 1) << "Invalid size of the sample target";
}

INSTANTIATE_TEST_SUITE_P(
  TestDatasetWithParams,
  TestInitializationParam,
  ::testing::Values("init", "corrupt", "preload"),
  [](const testing::TestParamInfo<std::string>& info)
  {
    return info.param;
  }
);

class TestDataset : public ::testing::Test
{
  protected:
    fs::path rootPath;
    long H, W;
    size_t datasetSize;
    AudioDataset* audioDataset = nullptr;

    void SetUp() override
    {
      datasetSize = 6;
      H = 1290;
      W = 13;
      MockDataset::Mock(rootPath, true, datasetSize, H, W);
      audioDataset = new AudioDataset(rootPath, nullptr);
    }

    void TearDown() override
    {
      fs::remove_all(rootPath);
      delete audioDataset;
      audioDataset = nullptr;
    }
};

TEST_F(TestDataset, classes)
{
  c10::Dict<int64_t, std::string> classesExpected;
  classesExpected.insert(0, "jazz");
  classesExpected.insert(1, "rock");

  c10::Dict<int64_t, std::string> classes = audioDataset->classes();
  ASSERT_EQ(classes.size(), classesExpected.size()) << "Invalid number of classes.";
  for (const auto& pair : classes)
  {
    ASSERT_TRUE(classesExpected.contains(pair.key())) << "Invalid class.";
    ASSERT_EQ(pair.value(), classesExpected.at(pair.key())) << "Invalid class.";
  }
}

TEST_F(TestDataset, size)
{
  torch::optional<size_t> size = audioDataset->size();
  ASSERT_EQ(size.value(), datasetSize) << "Invalid size of the dataset.";
}
